pkg/reference/reference.go

```diff
@@ -244,20 +252,23 @@ func (r *APIResolver) Resolve(ctx context.Context, req ResolutionRequest) (Resol
 // ResolveMultiple resolves the supplied MultiResolutionRequest. The returned
 // MultiResolutionResponse always contains valid values unless an error was
 // returned.
-func (r *APIResolver) ResolveMultiple(ctx context.Context, req MultiResolutionRequest) (MultiResolutionResponse, error) {
+func (r *APIResolver) ResolveMultiple(ctx context.Context, req MultiResolutionRequest) (MultiResolutionResponse, error) { //nolint:gocyclo
 	// Return early if from is being deleted, or the request is a no-op.
 	if meta.WasDeleted(r.from) || req.IsNoOp() {
 		return MultiResolutionResponse{ResolvedValues: req.CurrentValues, ResolvedReferences: req.References}, nil
 	}
 
 	// The references are already set - resolve them.
 	if len(req.References) > 0 {
-		vals := make([]string, len(req.References))
+		var vals []string
 		for i := range req.References {
 			if err := r.client.Get(ctx, types.NamespacedName{Name: req.References[i].Name}, req.To.Managed); err != nil {
 				return MultiResolutionResponse{}, errors.Wrap(err, errGetManaged)
 			}
-			vals[i] = req.Extract(req.To.Managed)
+			if meta.WasDeleted(req.To.Managed) {
+				continue
+			}
+			vals = append(vals, req.Extract(req.To.Managed))
```

Comment:
> For instance, if I reference 3 security groups, I want to make sure that they are all resolved correctly, not just the ones that are healthy.

We're not just talking about healthy here, but scheduled for deletion, pending finalizers. Somebody/something did "kubectl delete" on it. For many resources, the external deletion will happen regardless of being in use by something else, but for _some_ AWS resources that means they will not be deleted as long as they are in use (especially security groups[^sg])

This code would only apply if the reconcile policy is set to `Always`, so this feature will be "opt-in" anyway. I guess we could consider having another policy field/value for this if required, like `AlwaysPrune`, if we're worried about this.

One corner case I can think about is when you have an object with a "Block all access" policy, then you would want to delete the object before the policy, if possible. For security groups, the default would be to block so that fails safely, but for buckets the policy would be deleted before a non-empty bucket regardless of how we solve _this_ issue.

[^sg]:  As long as an interface or other security group uses/refers to it, it cannot be deleted. In the case where you have a server group refering to a client group (inbound), and another rule the other way, you're pretty stuck.